components/uniset/src/uniset.rs

```diff
@@ -0,0 +1,387 @@
+                capacity
+            )
+            .into());
+        }
+        if Some(parsed) <= prev {
+            return Err(format!(
+                "Serialization must be sorted. {:?} followed by {:?}",
+                Some(parsed),
+                prev
+            )
+            .into());
+        }
+        serialized_vec.push(parsed);
+        prev = Some(parsed);
+    }
+    if serialized_vec.len() % 2 != 0 {
+        return Err(format!(
+            "Serialization must be even. Serialization: {:?} Length: {}",
+            serialized_vec,
+            serialized_vec.len()
+        )
+        .into());
+    }
+    serialized_vec.shrink_to_fit(); // necessary if the length < capacity
+    Ok(serialized_vec)
+}
+
+/// UnicodeSet membership wrapper
+///
+/// Provides exposure to membership functions and constructors from serialized UnicodeSets
+/// and predefined ranges.
+/// Implements an inversion list.
+//#[derive(Copy, Clone, Debug, Eq)]
+pub struct UnicodeSet {
+    // If we wanted to use an array to keep the memory on the stack, there is an unsafe nightly feature
+    // https://doc.rust-lang.org/nightly/core/array/trait.FixedSizeArray.html
+    // Allows for traits of fixed size arrays
+    inv_list: Vec<u32>,
+}
+
+impl FromStr for UnicodeSet {
+    type Err = Box<dyn Error>;
+
+    fn from_str(serialize: &str) -> Result<Self, Self::Err> {
+        match parse_serial_string(serialize) {
+            Ok(serialize) => {
+                if serialize.len() % 2 != 0 {
+                    return Err("Array length must be even".into());
+                }
+                Ok(UnicodeSet {
+                    inv_list: serialize,
+                })
+            }
+            Err(e) => Err(e.into()),
+        }
+    }
+}
+
+impl UnicodeSet {
+    /// Returns Result of UnicodeSet from serialized string
+    ///
+    /// Returns an error if the serialized string fails to parse.
+    /// The serialized string requires starting capacity integer, followed by space delimited
+    /// integer code points. There must be an even number of elements (not including the
+    /// capacity int), and must be in ascending sorted order.
+    ///
+    /// Example String: `"4 0 5 10 15"` designates a capacity of size `4`, followed by 2 ranges
+    /// The ranges are `{0, 4}` and `{10, 14}` inclusive
+    pub fn new(serialize: &str) -> Result<UnicodeSet, Box<dyn Error>> {
+        match parse_serial_string(serialize) {
+            Ok(serialize) => Ok(UnicodeSet {
+                inv_list: serialize,
+            }),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Returns Result of UnicodeSet from a single pair of integers defining a range
+    ///
+    /// `start`: inclusive, `end`: exclusive
+    ///
+    /// Returns an error if the range is invalid (out of order and out of bounds).
+    ///
+    /// Example Call: `UnicodeSet::from_range(&0, &15)`
+    pub fn from_range(start: u32, end: u32) -> Result<UnicodeSet, Box<dyn Error>> {
```

Comment:
this is ok, although the `From` I suggested above is superior as it allows for:

```rust
let x: UnicodeSet = 0..15.into();
let x: UnicodeSet = 0..15.try_into()?;
```

Reply:
agreed and changed